test/app-luatest/protobuf_I32_test.lua

```diff
@@ -0,0 +1,276 @@
+local t = require('luatest')
+local ffi = require('ffi')
+local protobuf = require('protobuf')
+local g = t.group()
+
+g.test_float = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'float', 1}})
+    })
+    local result = protocol:encode('test', {val = 0.5})
+    t.assert_equals(string.hex(result), '0d0000003f')
+end
+
+g.test_float_min = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'float', 1}})
+    })
+    local result = protocol:encode('test', {val = -0x1.fffffep+127})
+    t.assert_equals(string.hex(result), '0dffff7fff')
+end
+
+g.test_float_max = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'float', 1}})
+    })
+    local result = protocol:encode('test', {val = 0x1.fffffep+127})
+    t.assert_equals(string.hex(result), '0dffff7f7f')
+end
+
+g.test_float_exception_type = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'float', 1}})
+    })
+    local msg = 'Field "val" of "float" type gets "string" type value.'
+    local data = {val = 'str'}
+    t.assert_error_msg_contains(msg, protocol.encode, protocol, 'test', data)
+end
+
+g.test_float_exception_NaN = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'float', 1}})
+    })
+    local msg = 'Input data for "val" field is NaN'
+    local data = {val = 0/0}
+    t.assert_error_msg_contains(msg, protocol.encode, protocol, 'test', data)
+end
+
+g.test_float_exception_inf = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'float', 1}})
+    })
+    local msg = 'Input data for "val" field is inf'
+    local data = {val = 1/0}
+    t.assert_error_msg_contains(msg, protocol.encode, protocol, 'test', data)
+end
+
+g.test_float_exception_cdata = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'float', 1}})
+    })
+    local msg = 'Field "val" of "float" type gets "cdata" type value.'
+    local data = {val = 15LL}
+    t.assert_error_msg_contains(msg, protocol.encode, protocol, 'test', data)
+end
+
+g.test_float_exception_size = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'float', 1}})
+    })
+    local msg = 'Input data for "val" field is "3.402823467e+38" and ' ..
+        'do not fit in "float"'
+    local data = {val = 0x1.fffffe018d3f8p+127}
+    t.assert_error_msg_contains(msg, protocol.encode, protocol, 'test', data)
+end
+
+g.test_fixed32_zero = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'fixed32', 1}})
+    })
+    local result = protocol:encode('test', {val = 0})
+    t.assert_equals(string.hex(result), '0d00000000')
+end
```

Comment:
If I undestand it right, the 'no presence' discipline means that a default value shouldn't be encoded. See [here](https://www.notion.so/tarantool/RFC-Protocol-Buffers-encoder-decoder-17ff2ff599974f9a9d0f39d9ab0a5284?pvs=25#75a59aafc71648009d5ab58daaed2f36) and [here](https://protobuf.dev/programming-guides/field_presence/). If I understand it right, the result should be empty.

The same for an empty string/bytes. And, please, verify floating-point and enum types too.

Also, please, add a test case that verifies that it is possible to distinguish a presence of a field with an empty message from a situation, when this field is missing. (Nested messages have 'explicit presence' discipline.)

Reply:
Fixed default values encoding and added tests for all numeric types, strings, bytes, enums and nested messages. To comply with explicit presence discipline now empty message is encoded as a string with zero length so it can be distinguished from missing field. For this purpose an `ex_presence` flag is added to `wireformat.encode_len` function, because messages are encoded as strings.

Reply:
I'm not sure that I understand what the 'explicit presence' means for a message. My guess is that it really doesn't matter, whether to say that it has 'no presence' or 'explicit presence' discipline, because it has no default value in fact. If a message field is missed, it is not encoded. If it is present (even as `{}`), it is encoded.

It seems, we don't need the `ex_presence` flag, because we never call `encode_len` with `ex_presence = true` if the value to encode is an empty sequence of bytes.

Also, it seems that we don't handle `box.NULL` as a field value. I propose to interpret it in the same way as `nil` on encoding.

Reply:
If we will encode `{}` for message without `ex_presence` as it was before it will not appear in resulting string. You said that in `explicit presence` we need to distinguish a presence of a field with an empty message from a situation, when this field is missing. 

For `encode_string` an empty string is a default value and it will not be encoded. If we have empty message in fact we have to encode an empty string because tag has to emerge in result for decoder to say that the message was empty but not missing. So `ex_presence` flag commands to `encode_string` that empty string is not a default value and it has to encode it.

As a result if a message field is not set (missing) we will get empty string because `encode_field` would not be called.
If a message field is set empty `{}` we will run `encode_field` and then `encode_string` with flag `ex_presence` and we will get tag and zero length in resulting string.

So the decoder will read the string and distinquish whether was the message field set empty or not even set.

Reply:
Got it, thanks for the explanation.

I'll open separate discussions regarding handling `box.NULL` values.